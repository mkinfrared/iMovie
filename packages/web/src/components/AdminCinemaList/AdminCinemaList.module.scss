.AdminCinemaList {
  display: grid;
  gap: 20px;
  grid-template-columns: repeat(3, 1fr);
  justify-content: space-evenly;
  justify-items: center;

  @media (max-width: 997px) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media (max-width: 670px) {
    grid-template-columns: repeat(1, 1fr);
  }
}
