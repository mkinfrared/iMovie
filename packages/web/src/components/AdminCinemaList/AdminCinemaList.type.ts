import { Dispatch } from "redux";

import { Cinema } from "store/reducers/cinemas/types";

export interface AdminCinemaListProps {
  dispatch: Dispatch;
  cinemas: Cinema[];
}
