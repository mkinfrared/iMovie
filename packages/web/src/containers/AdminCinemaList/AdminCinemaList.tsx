import React from "react";
import { useDispatch, useSelector } from "react-redux";

import AdminCinemaListFC from "components/AdminCinemaList";
import { selectCinemas } from "store/reducers/cinemas/selectors";

const AdminCinemaList = () => {
  const dispatch = useDispatch();
  const cinemas = useSelector(selectCinemas);

  return <AdminCinemaListFC dispatch={dispatch} cinemas={cinemas} />;
};

export default React.memo(AdminCinemaList);
