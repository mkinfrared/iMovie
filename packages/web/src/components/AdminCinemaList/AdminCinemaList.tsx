import React, { useCallback, useEffect, useMemo, useState } from "react";

import CinemaCard from "components/CinemaCard";
import { fetchCinemas } from "store/reducers/cinemas/actions";
import { loadableModal } from "utils/loadable";

import css from "./AdminCinemaList.module.scss";
import { AdminCinemaListProps } from "./AdminCinemaList.type";

const EditCinemaForm = loadableModal(() => import("containers/EditCinemaForm"));

const AdminCinemaList = ({ dispatch, cinemas }: AdminCinemaListProps) => {
  const [cinemaId, setCinemaId] = useState<number | null>(null);

  useEffect(() => {
    dispatch(fetchCinemas());
  }, [dispatch]);

  const handleEditClick = useCallback((id: number) => {
    setCinemaId(id);
  }, []);

  const closeForm = useCallback(() => {
    setCinemaId(null);
  }, []);

  const cards = useMemo(
    () =>
      cinemas.map(({ id, name, zipcode }) => (
        <CinemaCard
          key={id}
          zipcode={zipcode.code}
          cinemaId={id}
          cinemaName={name}
          cityName={zipcode.city.name}
          countryName={zipcode.country.name}
          stateName={zipcode.state.name}
          onEdit={handleEditClick}
        />
      )),
    [cinemas]
  );

  return (
    <div className={css.AdminCinemaList}>
      {cards}
      {cinemaId && (
        <EditCinemaForm id={cinemaId} onClose={closeForm} open={!!cinemaId} />
      )}
    </div>
  );
};

export default React.memo(AdminCinemaList);
