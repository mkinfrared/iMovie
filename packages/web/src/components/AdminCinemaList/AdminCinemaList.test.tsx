import { fireEvent, render, waitForElement } from "@testing-library/react";
import React, { ReactElement } from "react";
import { MemoryRouter } from "react-router-dom";

import { cinemaMock } from "store/reducers/cinemas/test/mocks";

import AdminCinemaList from "./AdminCinemaList";

jest.mock("containers/EditCinemaForm", () => (props: any) => (
  <div>
    <code>containers/EditCinemaForm</code>
    <code>{JSON.stringify(props)}</code>
  </div>
));

describe("<AdminCinemaList />", () => {
  const dispatch = jest.fn();
  const cinema = cinemaMock;

  let Component: ReactElement;

  beforeEach(() => {
    Component = (
      <MemoryRouter>
        <AdminCinemaList dispatch={dispatch} cinemas={[cinema]} />
      </MemoryRouter>
    );
  });

  it("should be defined", () => {
    const { container } = render(Component);

    expect(container).toBeDefined();
  });

  it("should match a snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should open 'EditCinemaForm'", async () => {
    const { queryByText, getByTestId } = render(Component);
    const editButton = getByTestId("editButton");

    let editCinemaForm = await queryByText("containers/EditCinemaForm");

    expect(editCinemaForm).not.toBeInTheDocument();

    fireEvent.click(editButton);

    await waitForElement(() => "containers/EditCinemaForm");

    editCinemaForm = await queryByText("containers/EditCinemaForm");

    expect(editCinemaForm).toBeInTheDocument();
  });
});
