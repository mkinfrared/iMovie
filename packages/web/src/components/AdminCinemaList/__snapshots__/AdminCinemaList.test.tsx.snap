// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminCinemaList /> should match a snapshot 1`] = `
<div>
  <div
    class="AdminCinemaList"
  >
    <div
      class="MuiPaper-root MuiCard-root CinemaCard MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5 MuiTypography-colorPrimary MuiTypography-gutterBottom MuiTypography-alignCenter"
        >
          Terrence
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
        >
          Details
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
        >
          City:
           
          <span
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Riverside
          </span>
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
        >
          State:
           
          <span
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            California
          </span>
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
        >
          Country:
           
          <span
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            United States of America
          </span>
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
        >
          Zipcode:
           
          <span
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            92506
          </span>
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          data-testid="editButton"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Edit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <a
          href="/admin/cinema/1"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              More
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
